### PR TITLE
Add FreeBSD support.

### PR DESCRIPTION
**Pull request by alexdupre**

### CODE REVIEW

**Comment by uranther** [comment]
Great work on the patch! I would like to see this merged to support BSD systems.

The metalinter test is producing warnings in webusb.go but everything else passes. The Travis CI setup should also include a FreeBSD box to test.

Are there any comments from the Trezor team? This PR is now over a month old.

File: usb/webusb.go [hunk 1]
```diff
@@ -216,7 +217,7 @@ func (b *WebUSB) match(dev lowlevel.Device) bool {
 	}
 	return (c.BNumInterfaces > webIfaceNum &&
 		c.Interface[webIfaceNum].Num_altsetting > webAltSetting &&
-		c.Interface[webIfaceNum].Altsetting[webAltSetting].BInterfaceClass == lowlevel.CLASS_VENDOR_SPEC)
+		(c.Interface[webIfaceNum].Altsetting[webAltSetting].BInterfaceClass == lowlevel.CLASS_VENDOR_SPEC || runtime.GOOS == "freebsd"))
```
Comment (prusnak) [inline usb/webusb.go]: please do:

```
if runtime.GOOS != "freebsd" {
    return X;
} else {
    return Y;
}
```

**Comment by prusnak** [comment]
Thanks, much better.

Can you elaborate why `lowlevel.Get_Port_Numbers()` does not work on FreeBSD as expected to have full picture?

**Comment by alexdupre** [comment]
I don't know if it was fixed recently in FreeBSD 11.2 by this commit https://svnweb.freebsd.org/base?view=revision&revision=331419 (`libusb_get_port_numbers` is a wrapper of `libusb20_dev_get_port_path`), but surely until 11.1 release that function wasn't returning the expected result.

**Comment by prusnak** [comment]
@alexdupre Do you plan to try 11.2 out? If you do, let's please wait until you try it out.

**Comment by alexdupre** [comment]
I'm going to do it, but not before 3 or 4 weeks.

**Comment by alexdupre** [comment]
It looks like the above change fixed the issue with FreeBSD 11.2. But FreeBSD 10.4 and 11.1 (that are not in EoL yet) are still affected.
What do you propose? I see at least the following two options:
- use the workaround for all freebsd versions
- remove the workaround and conditionally apply it in the port

**Comment by prusnak** [comment]
I would like to have minimal required changeset for FreeBSD support in our code. I propose the following:

Let's update the change to include only modifications needed to run on most recent FreeBSD (11.2).
Let's put the patch required for older versions of FreeBSD in the corresponding ports.

**Comment by alexdupre** [comment]
Done, for me it's ready to be merged, I've already conditionally added the libusb patches to the freebsd port.

**Comment by prusnak** [comment]
I merged some Windows fixes by @karel-3d recently, which resulted in conflicts in your PR.

I rebased/squashed your work on top of the current master, which is now available in this branch: https://github.com/trezor/trezord-go/tree/freebsd

Could you please confirm this works for you on FreeBSD before I merge this?

**Comment by alexdupre** [comment]
No, it's not building, let me check what's wrong.

**Comment by alexdupre** [comment]
The windows fixes seem too intrusive to me, an internal header file has been imported and a new field is added to an opaque struct and publicly accessed. No good. This breaks the compliance with any other libusb implementation.

**Comment by prusnak** [comment]
@alexdupre any ideas how to do it more cleanly?

**Comment by alexdupre** [comment]
> So what we need to do is to exclude devices without winusb drivers in device listing.
>
>I was too afraid to change the device listing in os/windows_winusb.c to exclude it directly there - because I was afraid of a memory leak (the memory is manually managed for the device objects).
>
>Instead, I did a dirtier solution 

I don't know if there is another solution (because I don't know the exact problem on windows), but modifying the winusb implementation to change the device listing and exclude it there is surely a cleaner / more correct solution.

**Comment by karel-3d** [comment]
Yeah I did it like this because the windows libusb code is just too wild and I was too afraid to touch its internals. I was afraid I will forget some `free` on device skipping and cause some memory leak, so I just added a new field and exposed it and filter on that level.

The sad thing is the new libusb version is a little cleaner on windows (but not that much), but that caused other issues to other windows users which were hard to find, so I had to revert to the old libusb (few months ago).

**Comment by alexdupre** [comment]
Instead of entirely skipping the device you might just slightly change the data to avoid it from being matched by the webusb match function. It'd be a hack, but still much better than the current one.

**Comment by prusnak** [comment]
> you might just slightly change the data to avoid it from being matched by the webusb match function

Maybe setting `num_configurations` or `attached` to 0?

**Comment by karel-3d** [comment]
Actually no we don't, I have added that now. Before it was not imported.

**Comment by karel-3d** [comment]
I will look at slightly changing the data, not a bad idea actually. I thought this was better because it is more explicit, I have not thought about compatibility with different libusb versions.

**Comment by karel-3d** [comment]
I have made an issue for that

https://github.com/trezor/trezord-go/issues/94

**Comment by karel-3d** [comment]
Now in https://github.com/trezor/trezord-go/pull/95

please try that

**Comment by prusnak** [comment]
I just merged #95. @alexdupre can you try rebasing your work on top of current master to see if the changes are inobtrusive enough for you?

**Comment by karel-3d** [comment]
I have rebased the branch to the current master, put that on `freebsd` branch here, and fixed the metalinter stuff. And I cleared the functions a bit:

1) HID layer does not need to be initialized
2) trezor 1 with webusb will use trezor 2 vid/pid in the end, so I simplified the matching function somewhat

I have tested that it builds on freebsd and that it correctly recognizes WebUSB device (through loading status page with `links` :D ), I have not yet tested t1 (with HID). I am now installing xorg and gnome on my freebsd machine to run firefox there :D

**Comment by karel-3d** [comment]
I have finally installed gnome on my freebsd machine :D 

I can confirm my rebased branch works on freebsd with both HID and WebUSB devices.

However. When you disconnect a HID device during communication, bridge stops listing any devices, and never lists anything. (Probably some lock somewhere.) This does not happen with WebUSB devices. I am not sure where is the issue, if it is with OS or with bridge, I will investigate more.

**Comment by alexdupre** [comment]
Thanks @karel-3d, tomorrow I'll try it myself and will try to reproduce the issue you mentioned. Does it happen with any trezor device when you disconnect it during usage?

**Comment by karel-3d** [comment]
I have solved the issue with Trezor 1 here (since it is more general)

https://github.com/trezor/trezord-go/pull/99

but I found one more issue with Trezor T, also with disconnecting during usage... I will try to replicate it

**Comment by karel-3d** [comment]
I cannot reproduce any issues now, on my FreeBSD machine. I am force-pushing my branch with additional fixes to your fork so we can merge this PR.

(Btw. I have, for fun, tried Chrome on FreeBSD and "native" WebUSB - it doesn't work with Trezor. Probably because FreeBSD is not in [vendored libusb in Chrome](https://github.com/chromium/chromium/tree/master/third_party/libusb/src/libusb/os), but who knows.)

**Comment by karel-3d** [comment]
I have force-pushed the commits to the other repo.

@prusnak , @alexdupre : if you both agree we can merge this. The disconnection issue is solved by the other PR.

**Comment by karel-3d** [comment]
> From the wiki I see it should work only on beta-wallet and only with Trezor T (that I haven't), is it true?

Trezor 1 fw on master now also has webusb.

**Comment by karel-3d** [comment]
Merging. Thanks for the patience.

**Comment by alexdupre** [comment]
Perfect, I'll update the FreeBSD port as soon as you tag a new release.

Regarding the Chrome WebUSB issue I've opened a bug report: https://github.com/gliaskos/freebsd-chromium/issues/132

**Comment by karel-3d** [comment]
Sorry for a little late reply, we had other priorities, but we are nearing a new release now. It's not tagged and internally tested yet but should be feature complete

**Comment by alexdupre** [comment]
Pushed to FreeBSD ports, thanks!

**Comment by prusnak** [comment]
Great, thanks!

**Pull request merged**